tests/subsys/input/api/src/main.c

```diff
@@ -0,0 +1,195 @@
+		message_count_filtered++;
+	}
+
+	k_sem_give(&cb_start);
+}
+INPUT_LISTENER_CB_DEFINE(&fake_dev, input_cb_filtered);
+
+static void input_cb_unfiltered(struct input_event *evt, bool sync)
+{
+	TC_PRINT("%s: %d\n", __func__, message_count_unfiltered);
+
+	message_count_unfiltered++;
+
+	if (message_count_unfiltered == CONFIG_INPUT_QUEUE_MAX_MSGS + 1) {
+		TC_PRINT("cb: done\n");
+		k_sem_give(&cb_done);
+	}
+}
+INPUT_LISTENER_CB_DEFINE(NULL, input_cb_unfiltered);
+
+ZTEST(input_api, test_sequence_thread)
+{
+	int i;
+	int ret;
+
+	message_count_filtered = 0;
+	message_count_unfiltered = 0;
+
+	k_sem_take(&cb_start, K_FOREVER);
+	k_sem_take(&cb_done, K_FOREVER);
+
+	/* fill the queue */
+	for (i = 0; i < CONFIG_INPUT_QUEUE_MAX_MSGS; i++) {
+		TC_PRINT("report: %d\n", i);
+		ret = input_report_key(&fake_dev, i, 1, false, K_FOREVER);
+		zassert_equal(ret, 0, "ret: %d", ret);
+	}
+
+	/* one extra with no dev to acconut for the message pending in the
+	 * locked cb
+	 */
+	ret = input_report_key(NULL, 0, 1, false, K_FOREVER);
+	zassert_equal(ret, 0, "ret: %d", ret);
+
+	zassert_false(input_queue_empty());
+
+	/* next message finds the queue full */
+	ret = input_report_key(&fake_dev, 0, 1, false, K_NO_WAIT);
+	zassert_equal(ret, -ENOMSG, "ret: %d", ret);
+
+	k_sem_give(&cb_start);
+
+	/* wait for cb to get all the messages */
+	k_sem_take(&cb_done, K_FOREVER);
+
+	zassert_equal(message_count_filtered, CONFIG_INPUT_QUEUE_MAX_MSGS);
+	zassert_equal(message_count_unfiltered, CONFIG_INPUT_QUEUE_MAX_MSGS + 1);
+}
+
+#else /* CONFIG_INPUT_MODE_THREAD */
+
+static void input_cb_filtered(struct input_event *evt, bool sync)
+{
+	if (evt->dev == &fake_dev) {
+		message_count_filtered++;
+	}
+}
+INPUT_LISTENER_CB_DEFINE(&fake_dev, input_cb_filtered);
+
+static void input_cb_unfiltered(struct input_event *evt, bool sync)
+{
+	message_count_unfiltered++;
+}
+INPUT_LISTENER_CB_DEFINE(NULL, input_cb_unfiltered);
+
+ZTEST(input_api, test_synchronous)
+{
+	int ret;
+
+	message_count_filtered = 0;
+	message_count_unfiltered = 0;
+
+	ret = input_report_key(&fake_dev, 0, 1, false, K_FOREVER);
+	zassert_equal(ret, 0, "ret: %d", ret);
+
+	ret = input_report_key(NULL, 0, 1, false, K_FOREVER);
+	zassert_equal(ret, 0, "ret: %d", ret);
+
+	zassert_equal(message_count_filtered, 1);
+	zassert_equal(message_count_unfiltered, 2);
+}
+
+static struct input_event last_event;
+static bool last_sync;
+
+static void input_cb_last_event(struct input_event *evt, bool sync)
+{
+	memcpy(&last_event, evt, sizeof(last_event));
+	last_sync = sync;
+}
+INPUT_LISTENER_CB_DEFINE(NULL, input_cb_last_event);
+
+ZTEST(input_api, test_report_apis)
+{
+	int ret;
+
+	ret = input_report_key(&fake_dev, INPUT_KEY_A, 1, false, K_FOREVER);
+	zassert_equal(ret, 0);
+	zassert_equal(last_event.dev, &fake_dev);
+	zassert_equal(last_event.type, INPUT_EV_KEY);
+	zassert_equal(last_event.code, INPUT_KEY_A);
+	zassert_equal(last_event.value, 1);
+	zassert_false(last_sync);
+
```

Comment:
It is necessary to clean up the `last_event` at this point. The next chunk of the test can result in a false positive because it is already populated.

Reply:
What? No that is not possible, the test events are unique, it would not hurt to do it but it's certainly not necessary. Please clarify, resolving the other comments.

Reply:
Use your are correct that would be too much. Maybe I am being conservative, making the test more “reliable” by not relying on the external code.